src/meep_internals.hpp

```diff
@@ -53,35 +54,36 @@ inline int rmin_bulk(int m) {
   return r;
 }
 
+// A source volume
+// Moveable and copyable
 class src_vol {
 public:
-  src_vol(component cc, src_time *st, size_t n, ptrdiff_t *ind, std::complex<double> *amps);
-  src_vol(const src_vol &sv);
-  ~src_vol() {
-    delete next;
-    delete[] index;
-    delete[] A;
-  }
-
-  src_time *t;
-  ptrdiff_t *index;        // list of locations of sources in grid (indices)
-  size_t npts;             // number of points in list
-  component c;             // field component the source applies to
-  std::complex<double> *A; // list of amplitudes
-
-  std::complex<double> dipole(size_t j) { return A[j] * t->dipole(); }
-  std::complex<double> current(size_t j) { return A[j] * t->current(); }
-  void update(double time, double dt) { t->update(time, dt); }
-
-  bool operator==(const src_vol &sv) const {
-    // note: don't compare sv.A, since this is used to see if we can just
-    // add one source's amplitudes to another in src_vol::add_to
-    return sv.npts == npts && sv.c == c && sv.t == t &&
-           memcmp(sv.index, index, npts * sizeof(ptrdiff_t)) == 0;
-  }
-
-  src_vol *add_to(src_vol *others);
-  src_vol *next;
+  // Constructs a new source volume. Takes ownership of `ind` and `amps`.
+  // Requirement: ind.size() == amps.size()
+  src_vol(component cc, src_time *st, std::vector<ptrdiff_t> &&ind,
+          std::vector<std::complex<double> > &&amps);
+
+  // Compares `a` and `b` for equality while ignoring the actual amplitudes.
+  static bool equal_except_for_amplitude(const src_vol &a, const src_vol &b);
```

Comment:
We could probably call this something like `coincide` or `combinable` since what it is really checking is that they coincide in space and field component, and it is used to check whether they can be combined into a single `src_vol` by adding amplitudes.

Reply:
Updated.